plugin/semisync/semisync_master.cc

```diff
@@ -554,10 +580,214 @@ bool ReplSemiSyncMaster::is_semi_sync_slave() {
+  char info_file_path[FN_REFLEN];
+  fn_format(tmp_file_path, SEMI_SYNC_MASTER_WHITELIST_TMP_FILE.c_str(),
+            mysql_data_home, "", MY_UNPACK_FILENAME | MY_RETURN_REAL_PATH);
+  fn_format(info_file_path, SEMI_SYNC_MASTER_WHITELIST_FILE.c_str(),
+            mysql_data_home, "", MY_UNPACK_FILENAME | MY_RETURN_REAL_PATH);
+
+  // case: whitelist file does not exists, we're done
+  if (my_access(info_file_path, F_OK)) {
+    return 0;
+  }
+
+  // case: temp file exists, delete it
+  if (!my_access(tmp_file_path, F_OK) &&
+      my_delete(tmp_file_path, MYF(MY_WME))) {
+    LogErr(ERROR_LEVEL, ER_SEMISYNC_FAILED_DELETE_TMP_WHITELIST_FILE,
+           tmp_file_path);
+    return 1;
+  }
+
+  std::ifstream info_file(info_file_path, std::ofstream::in);
+  if (!info_file.is_open()) {
+    LogErr(ERROR_LEVEL, ER_SEMISYNC_FAILED_TO_OPEN_WHITELIST_FILE,
+           info_file_path);
+    return 1;
+  }
+
+  std::string wlist, checksum_str;
+  // first line contains the whitelist
+  std::getline(info_file, wlist);
+  // second line constains the checksum
+  std::getline(info_file, checksum_str);
+#ifndef DEBUG_OFF
+  std::string tmp;
+  // assert: no more lines in the file
+  DBUG_ASSERT(!std::getline(info_file, tmp));
+#endif
+  info_file.close();
+
+  // verify the checksum
+  const ulong checksum = std::strtoul(checksum_str.c_str(), nullptr, 10);
+  const ulong checksum_calc =
+      crc32(0L, reinterpret_cast<const uchar *>(wlist.data()), wlist.length());
+
+  if (checksum != checksum_calc) {
+    LogErr(ERROR_LEVEL, ER_SEMISYNC_WHITELIST_CHECKSUM_DID_NOT_MATCH);
+    return 1;
+  }
+  // all good now, let's store the value in the data-structure and the sysvar
+  rpl_semi_sync_master_whitelist_set = split_into_set(wlist, ',');
+  // remove ANY from set
+  rpl_semi_sync_master_whitelist_set.erase(any_slave);
+  // store the value in the sysvar, hacky but works!
+  // NOTE: The plugin infrastructure will fill the default value (ANY) in the
+  // variable, so we have to free that first and then update it with what we
+  // found in the file
+  DBUG_ASSERT(rpl_semi_sync_master_whitelist &&
+              strcmp(rpl_semi_sync_master_whitelist, any_slave) == 0);
+  my_free(rpl_semi_sync_master_whitelist);
+  rpl_semi_sync_master_whitelist =
+      my_strdup(PSI_NOT_INSTRUMENTED, wlist.c_str(), MYF(MY_WME));
+
+  LogErr(INFORMATION_LEVEL, ER_SEMISYNC_WHITELIST_UPDATED, "",
+         rpl_semi_sync_master_whitelist);
+
+  return 0;
+}
+
+bool ReplSemiSyncMaster::update_whitelist(std::string &wlist) {
+  std::lock_guard<std::mutex> guard(rpl_semi_sync_master_whitelist_set_lock);
+
+  // remove all spaces
+  wlist.erase(std::remove(wlist.begin(), wlist.end(), ' '), wlist.end());
+
+  std::unordered_set<std::string> local_whitelist_set;
+  local_whitelist_set = rpl_semi_sync_master_whitelist_set;
+
+  // case: add a single uuid to the whitelist (value starts with +)
+  if (wlist[0] == '+') {
+    const auto str = wlist.substr(1);
```

Comment:
Should we also check that `str` is not a list.
Is `+<guid1>,<guid2>` syntax supported?

Reply:
no